packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,371 @@
+      propagateTraceHeaderUrls = [propagateTraceHeaderUrls];
+    }
+
+    const url = (span as tracing.Span).name;
+    const spanUrl = parseUrl(url);
+    if (spanUrl.origin !== window.location.origin) {
+      this._logger.warn('Cannot set headers on different origin');
+      return;
+    }
+
+    for (const propagateTraceHeaderUrl of propagateTraceHeaderUrls) {
+      if (sameOriginOrUrlMatches(url, propagateTraceHeaderUrl)) {
+        const b3Format = new B3Format();
+        const carrier: { [key: string]: unknown } = {};
+        b3Format.inject(span.context(), 'B3Format', carrier);
+
+        Object.keys(carrier).forEach(key => {
+          xhr.setRequestHeader(key, String(carrier[key]));
+        });
+      }
+    }
+  }
+
+  /**
+   * Adds span network events
+   * @param span
+   */
+  private _addSpanNetworkEvents(span: tracing.Span) {
+    const resources: PerformanceResourceTiming[] = otperformance.getEntriesByType(
+      'resource'
+    ) as PerformanceResourceTiming[];
+    const resource = getResource(
+      span,
+      resources,
+      this._usedResources[span.name]
+    );
+
+    if (resource) {
+      this._markResourceAsUsed(resource);
+      addSpanNetworkEvent(span, PTN.FETCH_START, resource);
+      addSpanNetworkEvent(span, PTN.DOMAIN_LOOKUP_START, resource);
+      addSpanNetworkEvent(span, PTN.DOMAIN_LOOKUP_END, resource);
+      addSpanNetworkEvent(span, PTN.CONNECT_START, resource);
+      addSpanNetworkEvent(span, PTN.SECURE_CONNECTION_START, resource);
+      addSpanNetworkEvent(span, PTN.CONNECT_END, resource);
+      addSpanNetworkEvent(span, PTN.REQUEST_START, resource);
+      addSpanNetworkEvent(span, PTN.RESPONSE_START, resource);
+      addSpanNetworkEvent(span, PTN.RESPONSE_END, resource);
+    }
+  }
+
+  /**
+   * Removes the previous information about span.
+   * This might happened when the same xhr is used again.
+   * @param xhr
+   * @private
+   */
+  private _cleanPreviousSpanInformation(xhr: XMLHttpRequestWrapped) {
+    const existingSpanId = xhr.__OT_SPAN_ID;
+    if (existingSpanId) {
+      const callbackToRemoveEvents = this._callbackToRemoveEvents[
+        existingSpanId
+      ];
+      if (callbackToRemoveEvents) {
+        callbackToRemoveEvents();
+      }
+    }
+  }
+
+  /**
+   * Creates a new span when method "open" is called
+   * @param xhr
+   * @param url
+   * @param method
+   * @private
+   */
+  private _createSpan(
+    xhr: XMLHttpRequestWrapped,
+    url: string,
+    method: string
+  ): tracing.Span | undefined {
+    if (isUrlIgnored(url, this._config.ignoreUrls)) {
+      this._logger.debug('ignoring span as url matches ignored url');
+      return;
+    }
+
+    const currentSpan = this._tracer.startSpan(url, {
+      parent: this._tracer.getCurrentSpan() || undefined,
+      attributes: {
+        [AttributeNames.COMPONENT]: this.component,
+        [AttributeNames.HTTP_METHOD]: method,
+        [AttributeNames.HTTP_URL]: url,
+      },
+    }) as tracing.Span;
```

Comment:
Why is a cast needed here? Can we remove it or modify the types to make it unnecessary?

Reply:
Yes it is needed, the thing is that general interface returns Span from types, but then basic tracer returns Span that extends Span and ReadableSpan. Then last thing is that `getCurrentSpan` from basic tracer returns Span which is then cast again as Span from types.

Reply:
OK, so you are suppressing the `ReadableSpan` part of the interface with this cast?

Could you instead use a type annotation, so `const currentSpan: tracing.Span = ...`. That's generally preferable to using `as` because it will correctly give a compiler error if the interface becomes incompatible, whereas `as` sort of forces the type more strongly.

Reply:
ts will also complain here

Reply:
What's the error message it gives? I wonder if we could clean the types up to allow them to be more consistent so it would work

Reply:
I still don't understand the need for the cast. `Span` does implement the `types.Span` interface so it shouldn't cause a type error